### PR TITLE
[Helm Chart] Only enable service monitor relabelings for prometheus scrape when prometheusServiceDiscovery is enabled

### PR DESCRIPTION
**Pull request by ufou**
Only enable service monitor relabelings for prometheus scrape when prometheusServiceDiscovery is enabled

**What type of PR is this?**
/kind bug

**What this PR does / Why we need it**:
This is a fix for the helm chart install. When the service monitor is used and `prometheusServiceDiscovery` is set to `false` (it's `true` by default) there is no need for the prometheus scrape relabelings on the service monitor resource, this PR ensures that the prometheus scrape relabelings are only enabled when `prometheusServiceDiscovery` is `true`

**Which issue(s) this PR fixes**:
None

**Special notes for your reviewer**:
I also had an issue with `install/helm/agones/templates/extensions-deployment.yaml` requiring the `mountPath` to be changed, but that may have been a side effect of using an incompatible image while testing:
```
         volumeMounts:
         - name: certs
-          mountPath: /certs
+          mountPath: /home/agones/certs
           readOnly: true
```

### CODE REVIEW

**Review by markmandel — COMMENTED** [review]
LGTM - @aimuz you implemented this change, just wanted to check with you before approving and merging?

**Review by markmandel — COMMENTED** [review]
Actually maybe a silly question - I don't know the Prometheus operator very well - if `.Values.agones.metrics.prometheusServiceDiscovery` is false, should we remove the `ServiceMonitor` entirely?

**Comment by aimuz** [comment]
I think the change needs to be tweaked. If prometheusServiceDiscovery is false, then no corresponding label will be created and ServiceMonitor will be useless.

If prometheusServiceDiscovery is false, I think the ServiceMonitor should not be created.

This is the current condition 

```
{{- if and (.Values.agones.metrics.prometheusEnabled) (.Values.agones.metrics.serviceMonitor.enabled) }}
```

to

```
{{- if and (.Values.agones.metrics.prometheusEnabled) (.Values.agones.metrics.serviceMonitor.enabled) (.Values.agones.metrics.prometheusServiceDiscovery) }}
```

File: install/helm/agones/templates/service-monitor.yaml [hunk 1]
```diff
@@ -35,6 +35,7 @@ spec:
       path: /metrics
       interval: {{ .Values.agones.metrics.serviceMonitor.interval }}
       relabelings:
+{{- if and (.Values.agones.metrics.prometheusServiceDiscovery) (.Values.agones.metrics.prometheusEnabled) }}
```
Comment (aimuz) [inline install/helm/agones/templates/service-monitor.yaml:38]:
```suggestion
{{- if (.Values.agones.metrics.prometheusServiceDiscovery) }}
```
prometheusEnabled is now always true

**Comment by ufou** [comment]
> prometheusServiceDiscovery

Perhaps my understanding of `prometheusServiceDiscovery` is incorrect in this chart? I always thought that if prometheus itself is discovering services then it is using the service annotations (so `prometheusServiceDiscovery` = true), whereas when prometheus operator is used the service annotations are not required because instead the service monitors

**Comment by aimuz** [comment]
You're right, I ignored the native prometheus who is using annotations to auto-discover the service.

What I'm not sure about now is that
prometheusServiceDiscovery = false
prometheus = true
ServiceMonitor = true

What would be the effect of ServiceMonitor? Is it going to work improperly?

If I remove the

```
        - action: labelmap
          regex: __meta_kubernetes_pod_label_(. +)
        - sourceLabels: [__meta_kubernetes_namespace]
          action: replace
          targetLabel: kubernetes_namespace
        - sourceLabels: [__meta_kubernetes_pod_name] action: replace targetLabel: kubernetes_namespace
          action: replace
          targetLabel: kubernetes_pod_name
```

Does ServiceMonitor work? If it works, will ServiceMonitor still work with prometheusServiceDiscovery = true?

**Comment by ufou** [comment]
> You're right, I ignored the native prometheus who is using annotations to auto-discover the service.
> 
> What I'm not sure about now is that prometheusServiceDiscovery = false prometheus = true ServiceMonitor = true
> 
> What would be the effect of ServiceMonitor? Is it going to work improperly?
> 
> If I remove the
> 
> ```
>         - action: labelmap
>           regex: __meta_kubernetes_pod_label_(. +)
>         - sourceLabels: [__meta_kubernetes_namespace]
>           action: replace
>           targetLabel: kubernetes_namespace
>         - sourceLabels: [__meta_kubernetes_pod_name] action: replace targetLabel: kubernetes_namespace
>           action: replace
>           targetLabel: kubernetes_pod_name
> ```
> 
> Does ServiceMonitor work? If it works, will ServiceMonitor still work with prometheusServiceDiscovery = true?

I think it's probably ok - I thought prometheus might end up with duplicated metrics as the serviceMonitor would essentially be adding the same service endpoints to prometheus as prometheus itself, I did wonder if the two options should be mutually exclusive, but it could just be that it's enough to say if `serviceMonitor` is enabled and `prometheusServiceDiscovery` is disabled, remove the relabelings

**Comment by ufou** [comment]
> > You're right, I ignored the native prometheus who is using annotations to auto-discover the service.
> > What I'm not sure about now is that prometheusServiceDiscovery = false prometheus = true ServiceMonitor = true
> > What would be the effect of ServiceMonitor? Is it going to work improperly?
> > If I remove the
> > ```
> >         - action: labelmap
> >           regex: __meta_kubernetes_pod_label_(. +)
> >         - sourceLabels: [__meta_kubernetes_namespace]
> >           action: replace
> >           targetLabel: kubernetes_namespace
> >         - sourceLabels: [__meta_kubernetes_pod_name] action: replace targetLabel: kubernetes_namespace
> >           action: replace
> >           targetLabel: kubernetes_pod_name
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does ServiceMonitor work? If it works, will ServiceMonitor still work with prometheusServiceDiscovery = true?
> 
> I think it's probably ok - I thought prometheus might end up with duplicated metrics as the serviceMonitor would essentially be adding the same service endpoints to prometheus as prometheus itself, I did wonder if the two options should be mutually exclusive, but it could just be that it's enough to say if `serviceMonitor` is enabled and `prometheusServiceDiscovery` is disabled, remove the relabelings

I mean, in general you either want to use a prometheus operator and service/pod monitors or prometheus with service annotations, not both

**Comment by aimuz** [comment]
I think it can just be removed without having to determine if prometheusServiceDiscovery is on.

https://github.com/googleforgames/agones/pull/3285/files#diff-6e33dcb8f0aab28e6040e6e7191d0b45df82f8f10a30278ff2bdeab87e09a04eR39-R50

```
       - sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
          action: keep
          regex: "true"
        - sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
          action: replace
          targetLabel: __metrics_path__
          regex: (.+)
        - sourceLabels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
          action: replace
          regex: ([^:]+)(?::\d+)?;(\d+)
          replacement: $1:$2
          targetLabel: __address__
```

I don't think this has any effect

**I apologize for posting the wrong code earlier.**

**Comment by aimuz** [comment]
At first it was thought that ServiceMonitor needed annotation to get the correct address, but now it seems that it doesn't, so I think this code can be removed without any impact.

So I think the final revision of this PR should look like this

```diff
       path: /metrics
       interval: {{ .Values.agones.metrics.serviceMonitor.interval }}
       relabelings:
-        - sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
-          action: keep
-          regex: "true"
-        - sourceLabels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
-          action: replace
-          targetLabel: __metrics_path__
-          regex: (.+)
-        - sourceLabels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
-          action: replace
-          regex: ([^:]+)(?::\d+)?;(\d+)
-          replacement: $1:$2
-          targetLabel: __address__
         - action: labelmap
           regex: __meta_kubernetes_pod_label_(.+)
         - sourceLabels: [__meta_kubernetes_namespace]
```

**Comment by markmandel** [comment]
Just triple checking with you @aimuz before approving this -- you clearly know this way more than I do 😄

**Review by aimuz — APPROVED** [review]
/lgtm I tested him and he's working.

**Comment by markmandel** [comment]
Awesome. Much appreciated, and apologies for the delay - was chasing down a gnarly bug. Approving and merging!

**Review by markmandel — APPROVED** [review]

**Comment by markmandel** [comment]
``
{"error":null,"message":"Allocate (public): failed allocation","result":{"metadata":{"name":"gsa-1","namespace":"default","creationTimestamp":null},"spec":{"multiClusterSetting":{"policySelector":{}},"required":{"gameServerState":"Ready"},"selectors":[{"gameServerState":"Ready"}],"scheduling":"Packed","metadata":{}},"status":{"state":"UnAllocated","gameServerName":"","source":""}},"severity":"info","test":"TestAllocatorAllocateOnGameServerUpdateError","time":"2023-07-31T23:56:04.228837136Z"}
--- FAIL: TestAllocatorAllocateOnGameServerUpdateError (25.23s)
    allocator_test.go:508: 
        	Error Trace:	/go/src/agones.dev/agones/pkg/gameserverallocations/allocator_test.go:508
        	Error:      	Expected nil, but got: &v1.GameServerAllocation{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"gsa-1", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.GameServerAllocationSpec{MultiClusterSetting:v1.MultiClusterSetting{Enabled:false, PolicySelector:v1.LabelSelector{MatchLabels:map[string]string(nil), MatchExpressions:[]v1.LabelSelectorRequirement(nil)}}, Required:v1.GameServerSelector{LabelSelector:v1.LabelSelector{MatchLabels:map[string]string(nil), MatchExpressions:[]v1.LabelSelectorRequirement(nil)}, GameServerState:(*v1.GameServerState)(0xc001bdf460), Players:(*v1.PlayerSelector)(nil), Counters:map[string]v1.CounterSelector{}, Lists:map[string]v1.ListSelector{}}, Preferred:[]v1.GameServerSelector(nil), Priorities:[]v1.Priority(nil), Selectors:[]v1.GameServerSelector{v1.GameServerSelector{LabelSelector:v1.LabelSelector{MatchLabels:map[string]string(nil), MatchExpressions:[]v1.LabelSelectorRequirement(nil)}, GameServerState:(*v1.GameServerState)(0xc001bdf470), Players:(*v1.PlayerSelector)(nil), Counters:map[string]v1.CounterSelector{}, Lists:map[string]v1.ListSelector{}}}, Scheduling:"Packed", MetaPatch:v1.MetaPatch{Labels:map[string]string(nil), Annotations:map[string]string(nil)}, Counters:map[string]v1.CounterAction(nil), Lists:map[string]v1.ListAction(nil)}, Status:v1.GameServerAllocationStatus{State:"UnAllocated", GameServerName:"", Ports:[]v1.GameServerStatusPort(nil), Address:"", NodeName:"", Source:"", Metadata:(*v1.GameServerMetadata)(nil)}}
        	Test:       	TestAllocatorAllocateOnGameServerUpdateError
```

This flake will be the death of me.

**Pull request merged**